File: output/copilot/python/hard/q38-trapping-rain-water.py
# https://leetcode.com/problems/trapping-rain-water/
# Problem: Trapping Rain Water

#Problem Description:
#Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.

# Constraints:
# n == height.length
# 1 <= n <= 2 * 104
# 0 <= height[i] <= 105

#Code Structure:
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if not height:
            return 0
        
        n = len(height)
        left_max = [0] * n
        right_max = [0] * n
        
        left_max[0] = height[0]
        for i in range(1, n):
            left_max[i] = max(left_max[i - 1], height[i])
        
        right_max[n - 1] = height[n - 1]
        for i in range(n - 2, -1, -1):
            right_max[i] = max(right_max[i + 1], height[i])
        
        water_trapped = 0
        for i in range(n):
            water_trapped += min(left_max[i], right_max[i]) - height[i]
        
        return water_trapped                               
    

File: output/copilot/php/hard/q38-trapping-rain-water.php
<?php
/** https://leetcode.com/problems/trapping-rain-water */
// Problem: Trapping Rain Water

//Problem Description:
//Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.

// Constraints:
// n == height.length
// 1 <= n <= 2 * 10^4
// 0 <= height[i] <= 10^5

//Code Structure:
class Solution {

    /**
     * Calculates how much water can be trapped after raining.
     * Uses two-pointer approach for O(n) time and O(1) space.
     * @param Integer[] $height
     * @return Integer
     */
    function trap($height) {
        $n = count($height);
        if ($n <= 2) return 0;

        $left = 0;
        $right = $n - 1;
        $leftMax = 0;
        $rightMax = 0;
        $water = 0;

        while ($left < $right) {
            if ($height[$left] < $height[$right]) {
                if ($height[$left] >= $leftMax) {
                    $leftMax = $height[$left];
                } else {
                    $water += $leftMax - $height[$left];
                }
                $left++;
            } else {
                if ($height[$right] >= $rightMax) {
                    $rightMax = $height[$right];
                } else {
                    $water += $rightMax - $height[$right];
                }
                $right--;
            }
        }

        return $water;
    }
}




